Inline the "body" of fields of modules when possible.

sjrd opened the pull request.

ekrich commented:
Does that speed things up too via the def to val changes?

sjrd commented:
The def to val change makes no difference per se since we now inline the body of the val where previously we inlined the body of the def.

Inlining the fields of `Predef` and `scala.package` can only improve performance, as well as not calling the module accessors that we don't need. I would be surprised if it can be *measured*, though.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -5140,6 +5205,41 @@ private[optimizer] object OptimizerCore {
+  }
+
+  object InlineableFieldBodies {
+    sealed abstract class FieldBody
```
I do not understand why we need this additional class structure. Aren't these the same as trees?

sjrd replied:
The idea here was to have something that we can reliably compare for equality, for the incremental invalidation. Normally we don't compare general Trees for equality.

gzm0 replied:
I see. IIUC, comparing trees for equality in general is safe, right? It's just expensive.

So having this additional structure doesn't seem to actually buy us anything in that regard; or am I missing something?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -1224,23 +1233,79 @@ private[optimizer] abstract class OptimizerCore(
-                RefinedType(expectedType)))
+    fieldBody match {
+      case FieldBody.Literal(literal) =>
+        // explicit copy for the position
```
Is this really what we want? For other inlining, IIUC we do not do this.

sjrd replied:
If we don't do it, we must take the `pos` of the trees into account when checking for equality of the `FieldBody`es. Otherwise an incremental run might not produce the same source map as a clean run.

sjrd replied:
OK, makes sense. I added positions in `FieldBody` to be able to use them.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:
```diff
@@ -709,7 +777,28 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
 
         case LoadModule(className) =>
-          dependenciesBuilder += className
+          isLoadModuleSideEffectFree(className)
```
Under compliant module init semantics, the first `LoadModule` returns `null`. It throws if they're `Fatal` and (as an implementation detail) stack-overflows under `Unchecked`.

The way the current code is written interprets a `LoadModule(self)` before `StoreModule` as a cyclic dependency on oneself. It seems to me that it's the best interpretation, but we could also make it directly non-side-effect-free.

gzm0 replied:
> It seems to me that it's the best interpretation

Agreed.

> we could also make it directly non-side-effect-free.

Nevertheless, I think I'd prefer this approach. It makes it explicit that a self load must be handled and how it is handled. I've completely missed this possibility in the review of #4938.

sjrd replied:
Another possibility would be to change the codegen to emit `This()` instead of `LoadModule(self)` when in the constructor and after the `StoreModule`. Then we wouldn't need to deal with `LoadModule(self)` any differently than other `LoadModule`s here. They would register a dependency on self, which the algorithm would then declare non elidable because cyclic. Fewer special cases.

gzm0 replied:
And then we'd disallow `LoadModule(self)` via `ClassDefChecker`? Seems much more complicated than adding an `if` here :-/

sjrd replied:
No that would still be allowed. It would just be less optimized.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:
```diff
@@ -748,17 +837,25 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
 
-        case StoreModule(_, _) => true
-        case _                 => isTriviallySideEffectFree(tree)
+        case StoreModule(_, _) =>
```
That's the assumption, yes. I thought `ClassDefChecker` actually checked that we're in the same class. In fact we should even be in the constructor of the same class.

It doesn't prevent inlining because we can never inline the constructor of a module class. The only legit caller of that constructor is the module accessor.

Shall I add the check in `ClassDefChecker`?

sjrd replied:
Huh, our current situation is worse than I thought. According to `ClassDefChecker` and `IRChecker`, `StoreModule` can store any target module with any source value that has the right type. That means in theory we could even have a `StoreModule(..., Null())` ... I have no idea why we designed it that way. At the same time `OptimizerCore` already assumes that `StoreModule(_, _)` is for the current class and with a `This()` rhs, when it tries to eliminate `StoreModule`s in `tryElimStoreModule`.

I'm tempted to simplify the whole thing to have only `StoreModule()` without any argument, with a deserialization hack.

gzm0 replied:
Resolved in #4943.

sjrd commented:
Updated with:

* Rebase on top of `main`
* Addressed simple comments
* New commit that changes the codegen to replace `LoadModule` by `This`
* Simplified computation of the Info based on the above: don't treat `LoadModule(self)` in a special way
* Replaced the progressive change of the immutable `Set` of dependencies by decrementing a mutable `Int`

I left the change of `StoreModule` representation and validation out of this PR. It is in #4943.

sjrd commented:
Rebased (without conflict) now that #4943 was merged.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:
```diff
@@ -851,6 +915,137 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
+
+          case Apply(_, LoadModule(moduleClassName), MethodIdent(methodName), Nil)
+              if !methodName.isReflectiveProxy =>
+            if (moduleClassName == Class.this.className) {
```
Same here: Do we need this branch?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:
```diff
@@ -851,6 +915,137 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
+            Some(FieldBody.LoadModule(moduleClassName))
+
+          case Select(LoadModule(moduleClassName), className, FieldIdent(fieldName)) =>
+            if (moduleClassName == Class.this.className) {
```
Do we still need this branch? IIUC if we get a self-load module now, we consider the class to have non-elidable ctors?

sjrd replied:
Good catch. We don't need it anymore. Nor the other one.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -1221,23 +1230,79 @@ private[optimizer] abstract class OptimizerCore(
+                  default
+                case Some(fieldBody) =>
+                  val qualSideEffects = checkNotNullStatement(preTransQual)
+                  val fieldBodyTree = fieldBodyToTree(fieldBody)
```
(non-actionable): It feels a bit odd that we construct an intermediate tree here only to pattern match again on it for pretransforming. But maybe that's a better option to keep the logic in one spot.

gzm0 reviewed (changes requested):
Nice! I agree that adjusting the codegen makes this easier to understand.

A couple of minor comments and a follow-up on the tree positioning.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:
```diff
@@ -664,6 +681,16 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
     }
 
+    def askInlineableFieldBodies(asker: Processable): OptimizerCore.InlineableFieldBodies = {
+      if (askHasElidableConstructors(asker)) {
```
Indeed, good point. I had to reorganize things a bit here to make it work.

sjrd commented:
Rebased and addressed the latest comments.
